src/combinatorics/young/young_diagram.lean

```diff
@@ -305,4 +305,100 @@ end
+begin
+  induction w generalizing c;
+  rw young_diagram.cells_of_row_lens,
+  { change false ↔ ∃ (h : _ < 0), _, simp },
+  { rcases c with ⟨⟨_, _⟩, _⟩,
+    { simp },
+    { simpa [w_ih, -finset.singleton_product, nat.succ_lt_succ_iff] } }
+end
+
+/-- Young diagram from a sorted list -/
+def of_row_lens (w : list ℕ) (hw : w.sorted (≥)) : young_diagram :=
+{ cells        := young_diagram.cells_of_row_lens w,
+  is_lower_set := begin
+    rintros ⟨i2, j2⟩ ⟨i1, j1⟩ ⟨hi : i1 ≤ i2, hj : j1 ≤ j2⟩ hcell,
+    rw [finset.mem_coe, young_diagram.mem_cells_of_row_lens] at hcell ⊢,
+    obtain ⟨h1, h2⟩ := hcell,
+    refine ⟨hi.trans_lt h1, _⟩,
+    calc j1 ≤ j2            : hj
+      ...   < w.nth_le i2 _ : h2
+      ...   ≤ w.nth_le i1 _ : _,
+    cases eq_or_lt_of_le hi,
+    { subst h },
+    { apply list.pairwise_iff_nth_le.mp hw _ _ _ h }
+  end }
+
+lemma mem_of_row_lens {w : list ℕ} {hw : w.sorted (≥)} {c : ℕ × ℕ} :
+  c ∈ of_row_lens w hw ↔ ∃ (h : c.fst < w.length), c.snd < w.nth_le c.fst h :=
+young_diagram.mem_cells_of_row_lens
+
+/-- The number of rows in `of_row_lens w hw` is the length of `w` -/
+lemma row_lens_length_of_row_lens {w : list ℕ} {hw : w.sorted (≥)} (hpos : ∀ x ∈ w, 0 < x) :
+  (of_row_lens w hw).row_lens.length = w.length :=
+begin
+  simp only [length_row_lens, col_len, nat.find_eq_iff, mem_cells, mem_of_row_lens,
+             lt_self_iff_false, is_empty.exists_iff, not_not],
+  exact ⟨id, λ n hn, ⟨hn, hpos _ (list.nth_le_mem _ _ hn)⟩⟩,
+end
+
+/-- The length of the `i`th row in `of_row_lens w hw` is the `i`th entry of `w` -/
+lemma row_len_of_row_lens {w : list ℕ} {hw : w.sorted (≥)}
+  (i : ℕ) (hi : i < w.length) : (of_row_lens w hw).row_len i = w.nth_le i hi :=
+by simp [row_len, nat.find_eq_iff, mem_of_row_lens, hi]
+
+/-- The left_inv direction of the equivalence -/
+lemma of_row_lens_to_row_lens_eq_self {μ : young_diagram} :
+  of_row_lens _ (row_lens_sorted μ) = μ :=
+begin
+  ext ⟨i, j⟩,
+  simp only [mem_cells, mem_of_row_lens, length_row_lens, nth_le_row_lens],
+  simpa [← mem_iff_lt_col_len, mem_iff_lt_row_len] using j.zero_le.trans_lt,
+end
+
+/-- The right_inv direction of the equivalence -/
+lemma row_lens_of_row_lens_eq_self {w : list ℕ} {hw : w.sorted (≥)} (hpos : ∀ x ∈ w, 0 < x) :
+  (of_row_lens w hw).row_lens = w :=
+begin
+  ext i r,
+  cases lt_or_ge i w.length,
+  { simp only [option.mem_def, ← list.nth_le_eq_iff, h, row_lens_length_of_row_lens hpos],
+    revert r,
+    simpa only [eq_iff_eq_cancel_right, nth_le_row_lens] using row_len_of_row_lens _ h },
+  { rw [list.nth_eq_none_iff.mpr h, list.nth_eq_none_iff.mpr],
+    rwa row_lens_length_of_row_lens hpos }
+end
+
+/-- Equivalence between Young diagrams and weakly decreasing lists of positive natural numbers.
+A Young diagram `μ` is equivalent to a list of row lengths. -/
+def equiv_list_row_lens : young_diagram ≃ {w : list ℕ // w.sorted (≥) ∧ ∀ x ∈ w, 0 < x} :=
```

Comment:
Could consider instead using
```suggestion
def equiv_list_row_lens : young_diagram ≃ {w : list ℕ+ // w.sorted (≥)} :=
```

Reply:
Hmm, my version of Lean doesn't recognize this notation. If I do `#check ℕ+` I get:
```
ℕ + ⁇ : Type
failed to synthesize type class instance for
⊢ has_add Type
```

Reply:
You need to import `pnat`

Reply:
Okay, I would rather not do this. It works OK in some places but it is going to require a lot of subtype bundling/unbundling (including going back and changing the definition of `row_lens` from #17061 and its associated lemmas).

Reply:
I tried making a separate equivalence at the end only:

```
def equiv_list_row_lens' : young_diagram ≃ {w : list ℕ+ // w.sorted (≥)} :=
equiv_list_row_lens.trans 
{ to_fun    := λ w, ⟨list.pmap subtype.mk w.1 w.2.2, 
                     list.pairwise.pmap w.2.1 w.2.2 (λ _ _ _ _, id)⟩,
  inv_fun   := λ w, ⟨list.map coe w.1, sorry, sorry⟩,
  left_inv  := sorry,
  right_inv := sorry,
}
```